src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.cs

```diff
@@ -81,7 +83,9 @@ private protected override void SetCreateObject(Delegate? createObject)
 
         internal JsonTypeInfo(JsonConverter converter, JsonSerializerOptions options)
             : base(typeof(T), converter, options)
-        { }
+        {
+            EffectiveConverter = converter is JsonConverter<T> jsonConverter ? jsonConverter : converter.CreateCastingConverter<T>();
```

Comment:
can we get this from PropertyInfoForTypeInfo?

Reply:
Possibly, but I think we should look at eventually removing `PropertyInfoForTypeInfo` in the future.